lib/menubar/init.lua

```diff
@@ -231,47 +231,74 @@ local function menulist_update(query, scr)
     -- beginning to the table shownitems, and the ones that contain
     -- command in the middle to the table match_inside.
 
+    local count_table = load_count_table()
+    local command_list = {}
+
+    local PRIO_NONE = 0
+    local PRIO_HIG = 3
+    local PRIO_LOW = 1
+    local PRIO_NORMAL = 2
+
     -- Add the categories
     if menubar.show_categories then
         for _, v in pairs(menubar.menu_gen.all_categories) do
             v.focused = false
             if not current_category and v.use then
                 if string.match(v.name, pattern) then
                     if string.match(v.name, "^" .. pattern) then
-                        table.insert(shownitems, v)
-                    else
-                        table.insert(match_inside, v)
+                        v.count = PRIO_NONE
+                        v.prio = PRIO_NORMAL
+
+                        -- use count from count_table if present
+                        if string.len(pattern) > 0 and count_table[v.name] ~= nil then
+                            v.count = tonumber(count_table[v.name])
+                        end
+
+                        if string.match(v.name, "^" .. pattern)
+                            or string.match(v.cmdline, "^" .. pattern) then
+                            v.prio = PRIO_HIGH
+                        else
+                            v.prio = PRIO_NORMAL
```

Comment:
This is dead code. There is an `if` above which makes sure that we only end up here when `"^" .. pattern` matches.

Also, this patch is wrong since it drops the `match_inside` case. This is the same problem that the original patch brought: We now only consider matches at the beginning of the category name and ignore "interior matches" (but the original patch introduced that problem for applications, not for categories).


How about introducing a helper function to avoid all this code duplication:
It begins with the `string.match` check above and ends with the `table.insert` below. It gets two arguments: The priority if we have a beginning-match and the priority for "interior matches".